great_expectations/expectations/metrics/map_metric_provider.py

```diff
@@ -2677,6 +2676,120 @@ def _spark_map_condition_rows(
     return filtered.limit(result_format["partial_unexpected_count"]).collect()
 
 
+def _spark_map_condition_index(
+    cls,
+    execution_engine: SparkDFExecutionEngine,
+    metric_domain_kwargs: Dict,
+    metric_value_kwargs: Dict,
+    metrics: Dict[str, Any],
+    **kwargs,
+) -> Union[List[Dict[str, Any]], None]:
+    """
+    Returns indices of the metric values which do not meet an expected Expectation condition for instances
+    of ColumnMapExpectation.
+
+    Requires `unexpected_index_column_names` to be part of `result_format` dict to specify primary_key columns
+    to return.
+    """
+    (
+        unexpected_condition,
+        compute_domain_kwargs,
+        accessor_domain_kwargs,
+    ) = metrics.get("unexpected_condition", (None, None, None))
+
+    if unexpected_condition is None:
+        return None
+
+    if "column" not in accessor_domain_kwargs:
+        raise ValueError(
+            """No "column" found in provided metric_domain_kwargs, but it is required for a column map metric
+(_spark_column_map_condition_values).
+"""
+        )
+    domain_kwargs = dict(**compute_domain_kwargs, **accessor_domain_kwargs)
+    domain_column = domain_kwargs["column"]
+
+    df: "pyspark.sql.dataframe.DataFrame" = execution_engine.get_domain_records(
+        domain_kwargs=domain_kwargs
+    )
+
+    result_format = metric_value_kwargs["result_format"]
+    if not result_format.get("unexpected_index_column_names"):
+        raise ge_exceptions.MetricResolutionError(
+            "unexpected_indices cannot be returned without 'unexpected_index_column_names'. Please check your configuration."
+        )
+
+    # withColumn is required to transform window functions returned by some metrics to boolean mask
+    data = df.withColumn("__unexpected", unexpected_condition)
+    filtered = data.filter(F.col("__unexpected") == True).drop(F.col("__unexpected"))
+    unexpected_index_list: Optional[List[Dict[str, Any]]] = []
+
+    unexpected_index_column_names: List[str] = result_format[
+        "unexpected_index_column_names"
+    ]
+    columns_to_keep: List[str] = [column for column in unexpected_index_column_names]
+    columns_to_keep.append(domain_column)
+
+    # check that column name is in row
+    for col_name in columns_to_keep:
+        if col_name not in filtered.columns:
+            raise ge_exceptions.InvalidMetricAccessorDomainKwargsKeyError(
+                f"Error: The unexpected_index_column '{col_name}' does not exist in Spark DataFrame. Please check your configuration and try again."
+            )
+
+    for row in filtered.collect():
+        dict_to_add: dict = {}
+        for col_name in columns_to_keep:
+            dict_to_add[col_name] = row[col_name]
+        unexpected_index_list.append(dict_to_add)
+
+    if result_format["result_format"] == "COMPLETE":
+        return unexpected_index_list
+    return unexpected_index_list[: result_format["partial_unexpected_count"]]
+
```

Comment:
Do you think we can do the filtering before the collect statement for the case where we are not returning the COMPLETE results? Otherwise it looks like here we are retrieving all of the unexpected values to the machine running GX and then filtering them if COMPLETE is not specified. I'm just concerned that this may cause undesired extra computation / transfer (especially for large datasets).

Reply:
It would be good to double check for this in the SQL case as well.

Reply:
great idea. I'll look into making the change both here and at the SQL stage

Reply:
This is already done for SQL